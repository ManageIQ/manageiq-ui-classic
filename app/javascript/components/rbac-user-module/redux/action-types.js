const prefix = '@@manageiq/users/';
export const LOAD_DATA = `${prefix}Load users data`;
export const SELECT_USERS = `${prefix}Select users from list`;
export const SAVE_USER = `${prefix}Save user`;
export const FETCH_DATA = `${prefix}Fetch data`;
export const FETCH_SUCESFULL = `${prefix}Fetch sucesfull`;
export const REQUEST_FAILED = `${prefix}Request failed`;
export const DELETE_USER = `${prefix}Delete user`;
export const LOAD_GROUPS = `${prefix}Load user groups`;
export const STORE_GROUPS = `${prefix}Store user groups`;
export const STORE_USERS_TREE = `${prefix}Store users tree`;
export const EDIT_USER = `${prefix}Edit user`;
export const STORE_TAG_CATEGORIES = `${prefix}Store tag categories`;
export const ADD_FLASH_MESSAGE = `${prefix}Add flash message`;
export const REMOVE_FLASH_MESSAGE = `${prefix}Remove flash message;`;
export const RESET_SELECTED_USERS = `${prefix}Reset selected users`;
export const STORE_USER_CUSTOM_EVENTS = `${prefix}Store user custom events`;
