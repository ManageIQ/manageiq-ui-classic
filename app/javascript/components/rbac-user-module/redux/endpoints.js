export const getUsersUrl = '/api/users?expand=resources&attributes=current_group,miq_groups,current_group.miq_user_role,tags&sort_by=name';
export const modifyUserUrl = userId => `/api/users${userId ? `/${userId}` : ''}`;
export const updateMiqUserTreeUrl = '/tree/ops_rbac?id=xx-u';
export const getUserGroupsUrl = '/api/groups?expand=resources';
export const getTagCategoriesUrl = '/ops/load_tags_categories';
export const updateUserTagsUrl = userId => `/api/users/${userId}/tags`;
