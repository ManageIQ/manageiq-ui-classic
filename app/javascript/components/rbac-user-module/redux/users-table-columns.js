export default [{
  property: 'name',
  label: 'Full Name',
}, {
  property: 'userid',
  label: 'Username',
}, {
  property: 'email',
  label: 'E-mail',
}, {
  property: 'current_group',
  label: 'Current Group',
}, {
  property: 'role',
  label: 'Role',
}, {
  property: 'lastlogon',
  label: 'Last Logon',
}, {
  property: 'lastlogoff',
  label: 'Last Logoff',
}];
