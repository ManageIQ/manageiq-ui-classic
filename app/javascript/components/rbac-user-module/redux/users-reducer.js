import * as actionTypes from './action-types';

const columns = [{
  property: 'name',
  label: 'Full Name',
}, {
  property: 'userid',
  label: 'Username',
}, {
  property: 'email',
  label: 'E-mail',
}, {
  property: 'current_group',
  label: 'Current Group',
}, {
  property: 'role',
  label: 'Role',
}, {
  property: 'lastlogon',
  label: 'Last Logon',
}, {
  property: 'lastlogoff',
  label: 'Last Logoff',
}];

const handleSelectUser = (users = [], user) => user.selected ? [ ...users, user] : users.filter(({ id }) => id !== user.id);

const usersReducer = (state = {
  isLoaded: false,
  isFetching: false,
  isValid: true,
  columns,
  flashMessages: [],
}, action) => {
  let newState = {};
  switch (action.type) {
    case actionTypes.FETCH_SUCESFULL:
      return { ...state, isFetching: false, isValid: true };
    case actionTypes.REQUEST_FAILED:
      return { ...state, isFetching: false, isValid: false };
    case actionTypes.FETCH_DATA:
    case actionTypes.SAVE_USER:
      return { ...state, isFetching: true };
    case actionTypes.LOAD_DATA:
      newState = {
        ...state,
        rows: [...action.data.rows],
        isLoaded: true,
        isFetching: false,
        isValid: true,
      };
      return { ...newState };
    case actionTypes.SELECT_USERS:
      return { ...state, selectedUsers: handleSelectUser(state.selectedUsers, action.selectedUser) };
    case actionTypes.RESET_SELECTED_USERS:
      return { ...state, selectedUsers: action.selectedUsers };
    case actionTypes.STORE_GROUPS:
      return { ...state, groups: action.groups };
    case actionTypes.STORE_USERS_TREE:
      return { ...state, usersTree: action.usersTree };
    case actionTypes.STORE_TAG_CATEGORIES:
      return { ...state, categories: action.categories };
    case actionTypes.ADD_FLASH_MESSAGE:
      return { ...state, flashMessages: [...state.flashMessages, action.flashMessage] }
    case actionTypes.REMOVE_FLASH_MESSAGE:
      return { ...state, flashMessages: state.flashMessages.filter(({ flashId }) =>  flashId !== action.flashMessage.flashId) }
    default:
      return { ...state };
  }
};
export default usersReducer;


