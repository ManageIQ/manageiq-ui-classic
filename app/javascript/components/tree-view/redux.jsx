/* eslint camelcase: ["warn", {allow: ["bs_tree", "tree_name", "click_url", "check_url", "allow_reselect", "hierarchical_check", "silent_activate", "select_node"]}] */
import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Tree, Node } from 'react-wooden-tree';

import { http } from '../../http_api';
import { combineReducers } from '../../helpers/redux';
import reducers, { ACTIONS } from './reducers/index';
import basicStore from './reducers/basicStore';
import {
  convert, callBack, activateNode,
} from './helpers';

const TreeView = (props) => {
  const {
    tree_name,
    bs_tree,
    check_url,
    click_url,
    oncheck,
    onclick,
    checkboxes,
    allow_reselect,
    hierarchical_check,
    callBack,
    silent_activate,
    select_node,
  } = props;

  const namespace = tree_name;
  const ConnectedNode = connect((store, ownProps) => ({ ...store[namespace][ownProps.nodeId] }))(Node);
  const ReduxTree = connect(store => ({ data: { ...store[namespace] } }))(Tree);

  /**
   * After the component mounts adds a tree specific reducer to the store
   */
  useEffect(() => {
    // Setting the URLs for the further actions:
    ManageIQ.tree.checkUrl = check_url;
    ManageIQ.tree.clickUrl = click_url;

    // FIXME: This check if the reducer exists is not ideal, but it is needed
    // when the tree is rerendered without page change (Edit Report Menu's).
    // When fixed the useEffect should remove the created reducer on unmount.
    if (!ManageIQ.redux.store.getState()[namespace]) {
      ManageIQ.redux.addReducer({
        [namespace]: combineReducers([
          basicStore,
          reducers(oncheck, onclick),
        ], namespace),
      });
    }
  }, []);

  /**
   * Populates the store from the prop by converting the supplied tree to
   * the correct format and then dispatching it to the store.
   */
  useEffect(() => {
    // FIXME - When the conversion wont be needed hopefuly in the future
    const tree = activateNode(convert(JSON.parse(bs_tree), node => node.state.checked, node => node.state.selected), silent_activate, select_node);

    callBack(null, ACTIONS.EMPTY_TREE, null, namespace);
    callBack(null, ACTIONS.ADD_NODES, tree, namespace);
  }, [bs_tree]);

  const onDataChange = commands => commands.forEach(command => callBack(command.nodeId, `@@tree/${command.type}`, command.value, namespace));

  /**
   * Lazy load function with a wrapper.
   * The lazy load should return promise but we are preprocessing the data
   * before passing to the component. For the indentitation to work we need to update the nodeId as
   * well, not only the key in the array.
   *
   * FIXME: Remove wrapper after server returning flat trees.
   */
  const lazyLoad = node => new Promise((resolve, reject) => {
    http.post(`/${ManageIQ.controller}/tree_autoload`, {
      id: node.attr.key,
      tree: tree_name,
      mode: 'all',
    }).then((result) => {
      const data = convert(result);
      let subtree = {};
      Object.keys(data).forEach((key) => {
        if (key !== '') {
          // Creating the node id from the parent id.
          const nodeId = `${node.nodeId}.${key}`;
          // Updating the children ids, so it does not point to something else.
          const element = { ...data[key], nodeId, nodes: data[key].nodes.map(child => `${node.nodeId}.${child}`) };
          subtree = { ...subtree, [nodeId]: element };
        }
      });
      resolve(subtree);
    }).catch(error => reject(error));
  });

  return (
    <ReduxTree
      nodeIcon=""
      expandIcon="fa fa-fw fa-angle-right"
      collapseIcon="fa fa-fw fa-angle-down"
      loadingIcon="fa fa-fw fa-spinner fa-pulse"
      checkedIcon="fa fa-fw fa-check-square-o"
      uncheckedIcon="fa fa-fw fa-square-o"
      selectedIcon=""
      partiallyCheckedIcon="fa fa-fw fa-check-square"
      connectedNode={ConnectedNode}
      checkable={checkboxes}
      showCheckbox={checkboxes}
      allowReselect={allow_reselect}
      callbacks={{ onDataChange, lazyLoad }}
      hierarchicalCheck={hierarchical_check}
      {...props}
    />
  );
};

TreeView.propTypes = {
  tree_name: PropTypes.string.isRequired,
  bs_tree: PropTypes.string.isRequired,
  checkboxes: PropTypes.bool,
  allow_reselect: PropTypes.bool,
  oncheck: PropTypes.string,
  onclick: PropTypes.string,
  check_url: PropTypes.string,
  click_url: PropTypes.string,
  callBack: PropTypes.func.isRequired,
  hierarchical_check: PropTypes.bool,
};

TreeView.defaultProps = {
  checkboxes: false,
  allow_reselect: false,
  oncheck: undefined,
  onclick: undefined,
  check_url: '',
  click_url: '',
  hierarchical_check: false,
};

const TreeViewRedux = connect(null, { callBack })(TreeView);

export default TreeViewRedux;
