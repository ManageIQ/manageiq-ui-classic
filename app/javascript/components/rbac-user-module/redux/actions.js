import { push } from 'connected-react-router';
import * as actionTypes from './action-types';
import * as endpoints from './endpoints';
import { API, http } from '../../../http_api';

const { sendDataWithRx } = window;

export const navigate = where => (dispatch, getState) => {
  const { pathname } = getState().router.location;
  if (pathname !== where) {
    dispatch(push(where));
  }
};

export const loadUsersData = data => ({
  type: actionTypes.LOAD_DATA,
  data,
});

export const storeUserGroups = groups => ({
  type: actionTypes.STORE_GROUPS,
  groups,
});

export const selectUsers = users => ({
  type: actionTypes.SELECT_USERS,
  selectedUsers: users,
});

const fetchData = type => ({
  type,
});

const fetchSucesfull = () => ({
  type: actionTypes.FETCH_SUCESFULL,
});

const fetchFailed = () => ({
  type: actionTypes.REQUEST_FAILED,
});

const storeUsersTree = usersTree => ({
  type: actionTypes.STORE_USERS_TREE,
  usersTree,
}); 

const disableTreeReactRouting = () => sendDataWithRx({ type: 'disable-react-routing' });
const handleUserDetailLinkAction = url => dispatch => {
  disableTreeReactRouting();
  dispatch(navigate('/'));
  miqOnClickSelectRbacTreeNode(url);
  ManageIQ.component.getComponentInstances('RbacModule')[0].destroy();
}

export const requestUsers = () => (dispatch) => {
  dispatch(fetchData(actionTypes.FETCH_DATA));
  return API.get(endpoints.getUsersUrl)
    .then(data => data.resources.map(item => ({
      ...item,
      current_group: {
        label: item.current_group.description,
        onClick: () => dispatch(handleUserDetailLinkAction(`g-${item.current_group.id}`)),
      },
      role: {
        label: item.current_group.miq_user_role.name,
        onClick: () => dispatch(handleUserDetailLinkAction(`ur-${item.current_group.miq_user_role.id}`)),
      },
      miq_groups: item.miq_groups.map(group => group.id),
      groups: item.miq_groups.map(group => ({
        label: group.description,
        icon: 'group',
        groupId: group.id,
        value: group.id,
        onClick: () => dispatch(handleUserDetailLinkAction(`g-${group.id}`)),
      })),
    })))
    .then(data => dispatch(loadUsersData({ rows: data })))
    .then(() => dispatch(fetchSucesfull()))
    .catch(() => dispatch(fetchFailed()));
};

export const handleUpdateUsersTree = () => dispatch =>
  http.get(endpoints.updateMiqUserTreeUrl)
    .then((response) => {
      sendDataWithRx({
        updateTreeNode: {
          tree: 'rbac_tree',
          key: 'xx-u',
          data: {
            nodes: [...response],
          },
        },
      });
      return dispatch(storeUsersTree(response));
    });

export const saveUser = user => (dispatch) => {
  dispatch(fetchData(actionTypes.SAVE_USER));
  return API.post(endpoints.modifyUserUrl(), user, { skipErrors: [400] })
    .then(
      () => dispatch(requestUsers()),
      (error) => { throw error })
    .then(() => dispatch(handleUpdateUsersTree()))
    .then(() => dispatch(createFlashMessage(sprintf(__('User "%s" was saved'), user.name), 'success')))
    .then(() => dispatch(navigate('/')))
    .catch((error) => {
      dispatch(createFlashMessage(error.data.error.message, 'error'));
      dispatch(fetchFailed);
    });
};

export const deleteUser = userId => (dispatch, getState) => {
  dispatch(fetchData(actionTypes.DELETE_USER));
  dispatch(navigate('/'));
  const name = getState().usersReducer.rows.find(({ id }) => id === userId);
  return API.delete(endpoints.modifyUserUrl(userId))
    .then(() => dispatch(createFlashMessage(sprintf(__('User "%s" was deleted'), name), 'success')))
    .then(() => dispatch(requestUsers()))
    .then(() => dispatch(handleUpdateUsersTree()))
    .catch((error) => {
      dispatch(fetchFailed());
      console.log(error);
    });
};

export const editUser = (user, userId) => (dispatch) => {
  dispatch(fetchData(actionTypes.EDIT_USER));
  return API.put(endpoints.modifyUserUrl(userId), user)
    .then(
      () => dispatch(navigate('/')),
      (err) => { throw err; },
    )
    .then(() => dispatch(createFlashMessage(sprintf(__('User "%s" was saved'), user.name), 'success')))
    .then(() => dispatch(requestUsers()))
    .then(() => dispatch(handleUpdateUsersTree()))
    .catch(() => dispatch(fetchFailed()));
};

export const loadGroups = () => (dispatch) => {
  dispatch(fetchData(actionTypes.LOAD_GROUPS));
  return API.get(endpoints.getUserGroupsUrl)
    .then(data => data.resources, (error) => { throw error; })
    .then(groups => groups.map(group => ({
      label: group.description,
      value: group.id,
    })))
    .then(groups => dispatch(storeUserGroups(groups)))
    .then(() => dispatch(fetchSucesfull()))
    .catch(() => dispatch(fetchFailed()));
};

const resetSelectedToolbarItems = () => sendDataWithRx({
  eventType: 'updateToolbarCount',
  countSelected: 0,
});

export const deleteMultipleusers = users => (dispatch) => {
  dispatch(fetchData(actionTypes.DELETE_USER));
  return API.post(endpoints.modifyUserUrl(), {
    action: 'delete',
    resources: users.map(user => ({ id: user.id })),
  })
    .then(() => users.map(({ name }) => dispatch(createFlashMessage(sprintf(__('User "%s" was deleted'), name), 'success'))))
    .then(
    () => dispatch(requestUsers()),
    (error) => { throw error; },
  )
    .then(() => dispatch(handleUpdateUsersTree()))
    .then(() => {
      resetSelectedToolbarItems();
      return dispatch(fetchSucesfull);
    })
    .catch(() => dispatch(fetchFailed()));
};

const sendTreeUpdate = (data) => {
  sendDataWithRx({
    updateTreeNode: {
      tree: 'rbac_tree',
      key: 'xx-u',
      data: {
        state: { selected: false },
        nodes: [...data],
      },
    },
  });
};

const markActiveUser = (users, key) => users.map(user => ({
  ...user,
  state: { selected: user.key === key },
}));

export const rowClicked = row => (dispatch, getState) => {
  const userKey = `u-${row.id}`;
  const tree = getState().usersReducer.usersTree;
  dispatch(navigate(`/preview/${row.id}`));
  if (!tree) {
    return http.get(endpoints.updateMiqUserTreeUrl)
      .then(users => markActiveUser(users, userKey))
      .then((data) => {
        dispatch(storeUsersTree(data));
        return data;
      })
      .then(data => sendTreeUpdate(data));
  }
  return sendTreeUpdate(markActiveUser(tree, userKey));
};

const storeTagCategories = categories => ({
  type: actionTypes.STORE_TAG_CATEGORIES,
  categories,
});
export const loadTagsCategories = () => dispatch => http.get(endpoints.getTagCategoriesUrl)
  .then(data => data.map(category =>
    ({ value: category.id, label: category.description, name: category.name })))
  .then(categories => categories.sort((a, b) => a.label.localeCompare(b.label)))
  .then(categories => dispatch(storeTagCategories(categories)));
/**
 * export const editUserTags = ({ unAssignPayloads, assignPayloads }) => (dispatch) => {
  dispatch(fetchData('update tags'));
  if (unAssignPayloads && unAssignPayloads.length > 0) {
    return Promise.all(unAssignPayloads.map(({ id, ...rest }) =>
      API.post(endpoints.updateUserTagsUrl(id), rest)))
      .then(() => { // eslint-disable-line
        if (assignPayloads && assignPayloads.length > 0) {
          return Promise.all(assignPayloads.map(({ id, ...rest }) =>
            API.post(endpoints.updateUserTagsUrl(id), rest)));
        }
      })
      .then(() => dispatch(requestUsers()))
      .then(() => dispatch(navigate('/')))
      .then(() => dispatch(fetchSucesfull));
  }
  if (assignPayloads && assignPayloads.length > 0) {
    return Promise.all(assignPayloads.map(({ id, ...rest }) =>
      API.post(endpoints.updateUserTagsUrl(id), rest)))
      .then(() => dispatch(requestUsers()))
      .then(() => dispatch(navigate('/')))
      .then(() => dispatch(fetchSucesfull));
  }
  return (dispatch(fetchFailed));
};
*/

export const createFlashMessage = (text, type) => ({
  type: actionTypes.ADD_FLASH_MESSAGE,
  flashMessage: {
    text,
    type,
    flashId: Date.now(),
  }
})

export const removeFlashMessage = flashMessage => ({
  type: actionTypes.REMOVE_FLASH_MESSAGE,
  flashMessage,
})
