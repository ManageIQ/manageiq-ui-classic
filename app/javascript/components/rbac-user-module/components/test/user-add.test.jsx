import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { ADD_FLASH_MESSAGE } from '../../redux/action-types';
import UserAdd from '../user-add';

describe('User add component', () => {
  const mockStore = configureStore([thunk]);
  let store;
  let filledStore;
  let initialProps;
  beforeEach(() => {
    store = mockStore({
      usersReducer: {},
    });
    initialProps = {
      match: { url: '', params: {} },
    };
    filledStore = mockStore({
      usersReducer: {
        rows: [{ id: '123', name: 'foo', groups: [{ groupId: '1' }] }],
        selectedUsers: [{
          id: '123',
          name: 'foo',
          userid: 'foo',
          groups: [{ groupId: '1' }],
        }],
        groups: [{
          value: '111',
          label: 'Group 111',
        }],
      },
    });
  });

  it('should render spinner', () => {
    const wrapper = shallow(<UserAdd store={store} {...initialProps} />).dive();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render form for user adding', () => {
    const wrapper = shallow(<UserAdd store={filledStore} {...initialProps} />).dive();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render form for user editing', () => {
    const wrapper = shallow(<UserAdd store={filledStore} {...initialProps} match={{ url: '/edit/123', params: { userId: '123' } }} />).dive();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render form form user copying', () => {
    const wrapper = shallow(<UserAdd store={filledStore} {...initialProps} match={{ url: '/add/copy', params: { copy: 'copy' } }} />).dive();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should dispatch correct actions when edit canceled', () => {
    const wrapper = shallow(<UserAdd store={filledStore} {...initialProps} />).dive();
    wrapper.instance().handleCancelClicked();
    const expectedActions = [{
      type: ADD_FLASH_MESSAGE,
      flashMessage: expect.objectContaining({
        type: 'info',
        flashId: expect.any(Number),
      }),
    }, {
      type: '@@router/CALL_HISTORY_METHOD',
      payload: expect.objectContaining({
        args: [],
        method: 'goBack',
      }),
    }];
    expect(filledStore.getActions()).toEqual(expectedActions);
  });
});
