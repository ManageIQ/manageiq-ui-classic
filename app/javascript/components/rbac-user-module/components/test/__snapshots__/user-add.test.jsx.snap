// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User add component should render form for user adding 1`] = `
<div>
  <h1>
    User add
  </h1>
  <RbacUserForm
    groups={
      Array [
        Object {
          "label": "Group 111",
          "value": "111",
        },
      ]
    }
    newRecord={true}
    onCancel={[Function]}
    onSave={[Function]}
  />
</div>
`;

exports[`User add component should render form for user editing 1`] = `
<div>
  <h1>
    User add
  </h1>
  <RbacUserForm
    groups={
      Array [
        Object {
          "label": "Group 111",
          "value": "111",
        },
      ]
    }
    initialValues={
      Object {
        "groups": Array [
          "1",
        ],
        "id": "123",
        "name": "foo",
      }
    }
    newRecord={false}
    onCancel={[Function]}
    onSave={[Function]}
  />
</div>
`;

exports[`User add component should render form form user copying 1`] = `
<div>
  <h1>
    User add
  </h1>
  <RbacUserForm
    groups={
      Array [
        Object {
          "label": "Group 111",
          "value": "111",
        },
      ]
    }
    initialValues={
      Object {
        "groups": Array [
          "1",
        ],
        "id": "123",
        "name": "foo",
        "userid": undefined,
      }
    }
    newRecord={true}
    onCancel={[Function]}
    onSave={[Function]}
  />
</div>
`;

exports[`User add component should render spinner 1`] = `
<div>
  <Spinner
    className=""
    inline={false}
    inverse={false}
    loading={true}
    size="lg"
  />
</div>
`;
